packages/ss58_codec/test/ss58_codec_test.dart

```diff
@@ -56,78 +56,178 @@ void main() {
+            '4uqUGom7PszVSaZipgHYwVNsMESj1W6cmMZYeXGFeeXGX6VEqm'),
+        throwsA(
+          predicate((exception) =>
+              exception is BadAddressLengthException &&
+              exception.toString() == 'Bad Length Address.'),
+        ),
+      );
+    });
+
+    /// example of invalid decoded address.
+    test('invalid check sum exception', () {
       expect(
-          () => SS58Codec.decode(
-              '4uqUGom7PszVSaZipgHYwVNsMESj1W6cmMZYeXGFeeXGX6VEqm'),
-          throwsException);
+        () => SS58Codec.decode(
+            '3HX1zEyzCbxeXe34JY8SNSVAZ6djFccsV5f67PTied4CcWHs'),
+        throwsA(
+          predicate((exception) =>
+              exception is InvalidCheckSumException &&
+              exception.toString() == 'Invalid checksum'),
+        ),
+      );
     });
   });
 
-  // Encode Exception testing
+  /// [SS58Codec] `encode` method Exception testing
   group('encode: throw exception -> ', () {
+    /// [Address] with invalid `prefix`
     test('invalid prefix ', () {
+      /// [Address] invalid because of `negative` prefix
+      final addressWithNegativePrefix =
+          Address(prefix: -1, bytes: Uint8List.fromList([]));
       expect(
-          () => SS58Codec.encode(
-              Address(prefix: -1, bytes: Uint8List.fromList([]))),
-          throwsA(isA<InvalidPrefixException>()));
+        () => SS58Codec.encode(addressWithNegativePrefix),
+        throwsA(
+          predicate((exception) =>
+              exception is InvalidPrefixException &&
+              exception.toString() == 'Invalid SS58 prefix byte: -1.'),
+        ),
+      );
+
+      /// [Address] invalid because of prefix is greater than the limit
+      final addressWithInvalidPrefix =
+          Address(prefix: 16384, bytes: Uint8List.fromList([]));
 
       expect(
-          () => SS58Codec.encode(
-              Address(prefix: 16384, bytes: Uint8List.fromList([]))),
-          throwsA(isA<InvalidPrefixException>()));
+        () => SS58Codec.encode(addressWithInvalidPrefix),
+        throwsA(
+          predicate((exception) =>
+              exception is InvalidPrefixException &&
+              exception.toString() == 'Invalid SS58 prefix byte: 16384.'),
+        ),
+      );
     });
 
+    /// [Address] with invalid `address.bytes.length`
     test('invalid address length at lengths: [3, 5, 6, 7]', () {
       for (var len in [3, 5, 6, 7]) {
+        final invalidAddress = Address(prefix: 0, bytes: Uint8List(len));
         expect(
-            () => SS58Codec.encode(Address(prefix: 0, bytes: Uint8List(len))),
-            throwsA(isA<BadAddressLengthException>()));
+          () => SS58Codec.encode(invalidAddress),
+          throwsA(
+            predicate(
+              (exception) =>
+                  exception is BadAddressLengthException &&
+                  exception.toString() ==
+                      'Bad Length Address: ${invalidAddress.toString()}.',
+            ),
+          ),
+        );
```

Comment:
nit: tests intrinsically have an expected value and an actual value. Notating those help with the test's readability. I'd also hard code `${invalidAddress.toString()}` to avoid changes on string conversion to affect the outcome of this test
```suggestion
         
          final expectedErrorMessage = 'Bad Length Address: ${invalidAddress.toString()}.' // TODO hardcode  ${invalidAddress.toString()}
          () => SS58Codec.encode(invalidAddress),
          throwsA(
            predicate(
              (exception) =>
                  exception is BadAddressLengthException &&
                  exception.toString() == expectedErrorMessage ,
            ),
          ),
        );
```

Reply:
I see, but as far as I know here its `pass by value` and `not pass by reference` so variable -> `invalidAddress` is never going to change itself or from within the encode function call.

Reply:
I agree that it's not going to change functionality, but we're writing tests. Readability and maintainability are really really super important to ensure that we don't have any bug in our tests (bc they are the source of truth deciding whether or not our code works. Finding bugs there is terrible) and they are not a pain to maintain. To the compiler it's the same, but it's a huge difference for us. As far as the compiler knows, this one-liner is good enough lol `() => SS58Codec.encode(invalidAddress),throwsA(predicate((exception) => exception is BadAddressLengthException && exception.toString() == 'Bad Length Address: ${invalidAddress.toString()}.' ,),),);`

Reply:
Alright haha, will try to maintain as much readability as possible.